### PR TITLE
reduce CPU in SortedMap and SortedSet #equals()

### PR DESCRIPTION
**Pull request by mkeskells**
When comparing sorted Map/Set, and the ordering is the same, we can iterate the collection, rather than lookups which are typically O(log(n))

Set when comparing a compatable SortedSet/Map the cpu used is O(n) not O(n * log(n))

For maps there is an allocation penalty, as each iterator generates tuples, and previously only one iterator was used

Benchmarks to follow

### CODE REVIEW

**Comment by mkeskells** [comment]
Mima issues
```
[error]  * abstract synthetic method scala$collection$immutable$SortedSet$$super=uals(java.lang.Object)Boolean in interface scala.collection.immutable.SortedSet is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.collection.immutable.SortedSet.scala$collection$immutable$SortedSet$$super=uals")
[error]  * abstract synthetic method scala$collection$immutable$SortedMap$$super=uals(java.lang.Object)Boolean in interface scala.collection.immutable.SortedMap is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.collection.immutable.SortedMap.scala$collection$immutable$SortedMap$$super=uals")
```

**Comment by NthPortal** [comment]
If you're enterprising, you could also optimise it for having the reverse `Ordering`s of each other (see `Ordering#isReverseOf`)

File: src/library/scala/collection/immutable/SortedMap.scala [hunk 1]
```diff
@@ -102,6 +102,20 @@ self =>
...
+    case _ if this eq that.asInstanceOf[AnyRef] => true
+    case sm: SortedMap[k, v] if sm.ordering == this.ordering =>
+      (sm canEqual this) &&
+        (this.size == sm.size) && {
+        val i1 = this.iterator
+        val i2 = sm.iterator
+        var allEqual = true
+        while (allEqual && i1.hasNext)
+          allEqual = i1.next() == i2.next()
+        allEqual
```
Comment (hrhino) [inline src/library/scala/collection/immutable/SortedMap.scala:115]: I would not personally object to
```scala
while (i1.hasNext) 
  if (i1.next() != i2.next())
    return false
true
```

**Review by hrhino — APPROVED** [review]

**Comment by mkeskells** [comment]
> If you're enterprising, you could also optimise it for having the reverse `Ordering`s of each other (see `Ordering#isReverseOf`)

It would be, but there isn't a `reverseIterator` available is there?

**Comment by NthPortal** [comment]
> It would be, but there isn't a `reverseIterator` available is there?

ah, so there's isn't. oh well

**Review by retronym — APPROVED** [review]

**Pull request merged**